File: hgs/test/test_Matrix.cpp
#include <gtest/gtest.h>

#include "Matrix.h"

TEST(TestMatrix, DimensionConstructors)
{
    Matrix<int> const square(10);
    ASSERT_EQ(square.size(), 10 * 10);
    ASSERT_EQ(square.max(), 0);  // matrix initialises all zero

    Matrix<int> const rectangle(10, 20);
    ASSERT_EQ(rectangle.size(), 10 * 20);
    ASSERT_EQ(square.max(), 0);  // matrix initialises all zero
}

TEST(TestMatrix, DataConstructor)
{
    std::vector<std::vector<int>> data = {{}, {}};
    Matrix<int> const empty(data);
    ASSERT_EQ(empty.size(), 0);

    Matrix<int> const nonEmpty({{1, 2}, {1, 2}});
    ASSERT_EQ(nonEmpty(0, 0), 1);
    ASSERT_EQ(nonEmpty(0, 1), 2);
}

TEST(TestMatrix, DataConstructorThrows)
{
    // Second row is shorter than first, so this should throw.
    ASSERT_THROW(Matrix<int>({{1, 2}, {1}}), std::invalid_argument);

    // Second row is longer than first, so this should also throw.
    ASSERT_THROW(Matrix<int>({{1, 2}, {1, 2, 3}}), std::invalid_argument);
}

TEST(TestMatrix, Elements)
{
    Matrix<int> m(10);

    for (auto i = 0; i != 10; ++i)
        for (auto j = 0; j != 10; ++j)
            m(i, j) = i + j;

    ASSERT_EQ(m.max(), 9 + 9);  // maximum value

    ASSERT_EQ(m(1, 1), 1 + 1);  // several elements
    ASSERT_EQ(m(2, 1), 2 + 1);
    ASSERT_EQ(m(1, 2), 1 + 2);
}

File: pyvrp/tests/test_Matrix.py
from numpy.testing import assert_equal, assert_raises

from pyvrp._pyvrp import Matrix


def test_dimension_constructor():
    square = Matrix(10)
    assert_equal(square.num_rows, 10)
    assert_equal(square.num_cols, 10)
    assert_equal(square.size(), 10 * 10)
    assert_equal(square.max(), 0)  # matrix initialises all zero

    rectangle = Matrix(10, 20)
    assert_equal(rectangle.num_rows, 10)
    assert_equal(rectangle.num_cols, 20)
    assert_equal(rectangle.size(), 10 * 20)
    assert_equal(rectangle.max(), 0)  # matrix initialises all zero


def test_data_constructor():
    empty = Matrix([[], []])
    assert_equal(empty.size(), 0)

    non_empty = Matrix([[1, 2], [1, 3]])
    assert_equal(non_empty.size(), 4)  # matrix has four elements
    assert_equal(non_empty.max(), 3)


def test_data_constructor_throws():
    with assert_raises(ValueError):
        Matrix([[1, 2], []])  # second row shorter than first

    with assert_raises(ValueError):
        Matrix([[1, 2], [1, 2, 3]])  # second row longer than first


def test_element_access():
    mat = Matrix(10)

    for i in range(10):
        for j in range(10):
            mat[i, j] = i + j

    assert_equal(mat.max(), 9 + 9)  # maximum value

    assert_equal(mat[1, 1], 1 + 1)  # several elements
    assert_equal(mat[2, 1], 2 + 1)
    assert_equal(mat[1, 2], 1 + 2)
